flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/BlocklistDeclarativeSlotPoolTest.java

```diff
@@ -0,0 +1,239 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.runtime.jobmaster.slotpool.DefaultDeclarativeSlotPoolTest.FreeSlotConsumer;
+import static org.apache.flink.runtime.jobmaster.slotpool.DefaultDeclarativeSlotPoolTest.NewSlotsService;
+import static org.apache.flink.runtime.jobmaster.slotpool.DefaultDeclarativeSlotPoolTest.createResourceRequirements;
+import static org.apache.flink.runtime.jobmaster.slotpool.DefaultDeclarativeSlotPoolTest.createSlotOffersForResourceRequirements;
+import static org.apache.flink.runtime.jobmaster.slotpool.DefaultDeclarativeSlotPoolTest.drainNewSlotService;
+import static org.apache.flink.runtime.jobmaster.slotpool.DefaultDeclarativeSlotPoolTest.increaseRequirementsAndOfferSlotsToSlotPool;
+import static org.apache.flink.shaded.guava30.com.google.common.collect.Iterables.getOnlyElement;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.HamcrestCondition.matching;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+
+/** Test for {@link BlocklistDeclarativeSlotPool}. */
+class BlocklistDeclarativeSlotPoolTest {
+
+    private static final ResourceProfile RESOURCE_PROFILE =
+            ResourceProfile.newBuilder().setCpuCores(1.7).build();
+
+    @Test
+    void testOfferSlotsFromBlockedTaskManager() throws Exception {
+        testOfferSlots(true);
+    }
+
+    @Test
+    void testOfferSlotsFromUnblockedTaskManager() throws Exception {
+        testOfferSlots(false);
+    }
+
+    private void testOfferSlots(boolean isBlocked) throws Exception {
+        final TaskManagerLocation taskManager = new LocalTaskManagerLocation();
+
+        final NewSlotsService notifyNewSlots = new NewSlotsService();
+        // mark task manager as blocked.
+        final BlocklistDeclarativeSlotPool slotPool =
+                BlocklistDeclarativeSlotPoolBuilder.builder()
+                        .setBlockedTaskManagerChecker(
+                                isBlocked ? taskManager.getResourceID()::equals : ignore -> false)
+                        .build();
+        slotPool.registerNewSlotsListener(notifyNewSlots);
+
+        final ResourceCounter resourceRequirements = createResourceRequirements();
+        slotPool.increaseResourceRequirementsBy(resourceRequirements);
+
+        // offer slots on the blocked task manager
+        Collection<SlotOffer> slotOffers =
+                createSlotOffersForResourceRequirements(resourceRequirements);
+
+        if (isBlocked) {
+            assertThat(SlotPoolTestUtils.offerSlots(slotPool, slotOffers, taskManager)).isEmpty();
+            assertThat(drainNewSlotService(notifyNewSlots)).isEmpty();
+        } else {
+            assertThat(SlotPoolTestUtils.offerSlots(slotPool, slotOffers, taskManager))
+                    .containsExactlyInAnyOrderElementsOf(slotOffers);
+            assertThat(drainNewSlotService(notifyNewSlots))
+                    .is(
+                            matching(
```

Comment:
I would avoid use Hamcrest in new tests. 
One possible way to do the verification is to build a `Map<AllocationID, SlotOffer> slotOfferMap` from `slotOffers`, and introduce a method `boolean matchSlotToOffers(PhysicalSlot, SlotOffer)` and the assertion can be:
```
assertThat(drainNewSlotService(notifyNewSlots)).allMatch(slot -> matchSlotToOffers(slot, slotOfferMap.remove(slot.getAllocationId()));
```